src/note.ts

```diff
@@ -13,15 +13,254 @@
+  modLeftPx: number;
+  /** Start `X` for right modifiers. */
+  modRightPx: number;
+  /** Extra space on left of note. */
+  leftDisplacedHeadPx: number;
+  glyphPx?: number;
+  /** Extra space on right of note. */
+  rightDisplacedHeadPx: number;
+}
+
+export interface NoteDuration {
+  duration: string;
+  dots: number;
+  type: string;
+}
+
+export interface DurationCode {
+  common: Type;
+  type: Record<string, Type>;
+}
+
+export interface Type extends KeyProps {
+  getWidth(scale?: number): number;
+
+  code: string;
+  code_head: string;
+  stem: boolean;
+  rest: boolean;
+  flag: boolean;
+  stem_offset: number;
+  stem_up_extension: number;
+  stem_down_extension: number;
+  tabnote_stem_up_extension: number;
+  tabnote_stem_down_extension: number;
+  dot_shiftY: number;
+  line_above: number;
+  line_below: number;
+  beam_count: number;
+  code_flag_upstem: string;
+  code_flag_downstem: string;
+  position: string;
+}
+
+export interface NameValue {
+  name: string;
+}
+
+export interface NoteRenderOptions {
+  draw_stem_through_stave?: boolean;
+  draw_dots?: boolean;
+  draw_stem?: boolean;
+  y_shift?: number;
+  extend_left?: number;
+  extend_right?: number;
+  glyph_font_scale?: number;
+  annotation_spacing: number;
+  glyph_font_size?: number;
+  scale?: number;
+  font?: string;
+  stroke_px?: number;
+}
+
+export interface ParsedNote {
+  duration: string;
+  type: string;
+  customTypes: string[];
+  dots: number;
+  ticks: number;
+}
+
+export interface KeyProps {
+  stem_down_x_offset: number;
+  stem_up_x_offset: number;
+  key: string;
+  octave: number;
+  line: number;
+  int_value: number;
+  accidental: string;
+  code: string;
+  stroke: number;
+  shift_right: number;
+  displaced: boolean;
+}
 
-export class Note extends Tickable {
-  static get CATEGORY() {
+export interface NoteStruct {
+  line: number;
+  /** The number of dots, which affects the duration. */
+  dots: number;
+  keys: string[];
+  /** The note type (e.g., `r` for rest, `s` for slash notes, etc.). */
+  type: string;
+  align_center: boolean;
+  duration_override: Fraction;
+  /** The time length (e.g., `q` for quarter, `h` for half, `8` for eighth etc.). */
+  duration: string;
+}
+
+export interface TabNotePositon {
+  fret: string;
+  str: number;
+}
+
+/**
+ * Note implements an abstract interface for notes and chords that
+ * are rendered on a stave. Notes have some common properties: All of them
+ * have a value (e.g., pitch, fret, etc.) and a duration (quarter, half, etc.)
+ *
+ * Some notes have stems, heads, dots, etc. Most notational elements that
+ * surround a note are called *modifiers*, and every note has an associated
+ * array of them. All notes also have a rendering context and belong to a stave.
+ */
+export abstract class Note extends Tickable {
+  protected stave?: Stave;
+  protected render_options: NoteRenderOptions;
+  protected duration: string;
+  protected positions?: TabNotePositon[];
```

Comment:
Seems like we could actually move this to `TabNote`. It's not used anywhere here, no?

Reply:
deleted